File: api/src/test/java/org/apache/iceberg/transforms/TestProjection.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iceberg.transforms;

import static org.apache.iceberg.TestHelpers.assertAndUnwrap;
import static org.apache.iceberg.TestHelpers.assertAndUnwrapUnbound;
import static org.apache.iceberg.expressions.Expressions.and;
import static org.apache.iceberg.expressions.Expressions.bucket;
import static org.apache.iceberg.expressions.Expressions.day;
import static org.apache.iceberg.expressions.Expressions.equal;
import static org.apache.iceberg.expressions.Expressions.greaterThanOrEqual;
import static org.apache.iceberg.expressions.Expressions.hour;
import static org.apache.iceberg.expressions.Expressions.lessThanOrEqual;
import static org.apache.iceberg.expressions.Expressions.month;
import static org.apache.iceberg.expressions.Expressions.or;
import static org.apache.iceberg.expressions.Expressions.truncate;
import static org.apache.iceberg.expressions.Expressions.year;
import static org.apache.iceberg.types.Types.NestedField.optional;
import static org.apache.iceberg.types.Types.NestedField.required;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.apache.iceberg.PartitionSpec;
import org.apache.iceberg.Schema;
import org.apache.iceberg.exceptions.ValidationException;
import org.apache.iceberg.expressions.BoundPredicate;
import org.apache.iceberg.expressions.Expression;
import org.apache.iceberg.expressions.Expressions;
import org.apache.iceberg.expressions.Or;
import org.apache.iceberg.expressions.Projections;
import org.apache.iceberg.expressions.UnboundPredicate;
import org.apache.iceberg.relocated.com.google.common.collect.Lists;
import org.apache.iceberg.types.Types;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

public class TestProjection {
  private static final Schema SCHEMA = new Schema(optional(16, "id", Types.LongType.get()));

  @Test
  public void testIdentityProjection() {
    List<UnboundPredicate<?>> predicates =
        Lists.newArrayList(
            Expressions.notNull("id"),
            Expressions.isNull("id"),
            Expressions.lessThan("id", 100),
            Expressions.lessThanOrEqual("id", 101),
            Expressions.greaterThan("id", 102),
            Expressions.greaterThanOrEqual("id", 103),
            Expressions.equal("id", 104),
            Expressions.notEqual("id", 105));

    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).identity("id").build();

    for (UnboundPredicate<?> predicate : predicates) {
      // get the projected predicate
      Expression expr = Projections.inclusive(spec).project(predicate);
      UnboundPredicate<?> projected = assertAndUnwrapUnbound(expr);

      // check inclusive the bound predicate to ensure the types are correct
      BoundPredicate<?> bound = assertAndUnwrap(predicate.bind(spec.schema().asStruct(), true));

      assertThat(projected.ref().name())
          .as("Field name should match partition struct field")
          .isEqualTo("id");
      assertThat(projected.op()).isEqualTo(bound.op());

      if (bound.isLiteralPredicate()) {
        assertThat(projected.literal().value())
            .isEqualTo(bound.asLiteralPredicate().literal().value());
      } else {
        assertThat(projected.literal()).isNull();
      }
    }
  }

  @Test
  public void testCaseInsensitiveIdentityProjection() {
    List<UnboundPredicate<?>> predicates =
        Lists.newArrayList(
            Expressions.notNull("ID"),
            Expressions.isNull("ID"),
            Expressions.lessThan("ID", 100),
            Expressions.lessThanOrEqual("ID", 101),
            Expressions.greaterThan("ID", 102),
            Expressions.greaterThanOrEqual("ID", 103),
            Expressions.equal("ID", 104),
            Expressions.notEqual("ID", 105));

    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).identity("id").build();

    for (UnboundPredicate<?> predicate : predicates) {
      // get the projected predicate
      Expression expr = Projections.inclusive(spec, false).project(predicate);
      UnboundPredicate<?> projected = assertAndUnwrapUnbound(expr);

      // check inclusive the bound predicate to ensure the types are correct
      BoundPredicate<?> bound = assertAndUnwrap(predicate.bind(spec.schema().asStruct(), false));

      assertThat(projected.ref().name())
          .as("Field name should match partition struct field")
          .isEqualTo("id");
      assertThat(projected.op()).isEqualTo(bound.op());

      if (bound.isLiteralPredicate()) {
        assertThat(projected.literal().value())
            .isEqualTo(bound.asLiteralPredicate().literal().value());
      } else {
        assertThat(projected.literal()).isNull();
      }
    }
  }

  @Test
  public void testCaseSensitiveIdentityProjection() {
    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).identity("id").build();
    Assertions.assertThatThrownBy(
            () -> Projections.inclusive(spec, true).project(Expressions.notNull("ID")))
        .isInstanceOf(ValidationException.class)
        .hasMessageContaining("Cannot find field 'ID' in struct");
  }

  @Test
  public void testStrictIdentityProjection() {
    List<UnboundPredicate<?>> predicates =
        Lists.newArrayList(
            Expressions.notNull("id"),
            Expressions.isNull("id"),
            Expressions.lessThan("id", 100),
            Expressions.lessThanOrEqual("id", 101),
            Expressions.greaterThan("id", 102),
            Expressions.greaterThanOrEqual("id", 103),
            Expressions.equal("id", 104),
            Expressions.notEqual("id", 105));

    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).identity("id").build();

    for (UnboundPredicate<?> predicate : predicates) {
      // get the projected predicate
      Expression expr = Projections.strict(spec).project(predicate);
      UnboundPredicate<?> projected = assertAndUnwrapUnbound(expr);

      // check inclusive the bound predicate to ensure the types are correct
      BoundPredicate<?> bound = assertAndUnwrap(predicate.bind(spec.schema().asStruct(), true));

      assertThat(projected.ref().name())
          .as("Field name should match partition struct field")
          .isEqualTo("id");
      assertThat(projected.op()).isEqualTo(bound.op());

      if (bound.isLiteralPredicate()) {
        assertThat(projected.literal().value())
            .isEqualTo(bound.asLiteralPredicate().literal().value());
      } else {
        assertThat(projected.literal()).isNull();
      }
    }
  }

  @Test
  public void testCaseInsensitiveStrictIdentityProjection() {
    List<UnboundPredicate<?>> predicates =
        Lists.newArrayList(
            Expressions.notNull("ID"),
            Expressions.isNull("ID"),
            Expressions.lessThan("ID", 100),
            Expressions.lessThanOrEqual("ID", 101),
            Expressions.greaterThan("ID", 102),
            Expressions.greaterThanOrEqual("ID", 103),
            Expressions.equal("ID", 104),
            Expressions.notEqual("ID", 105));

    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).identity("id").build();

    for (UnboundPredicate<?> predicate : predicates) {
      // get the projected predicate
      Expression expr = Projections.strict(spec, false).project(predicate);
      UnboundPredicate<?> projected = assertAndUnwrapUnbound(expr);

      // check inclusive the bound predicate to ensure the types are correct
      BoundPredicate<?> bound = assertAndUnwrap(predicate.bind(spec.schema().asStruct(), false));

      assertThat(projected.ref().name())
          .as("Field name should match partition struct field")
          .isEqualTo("id");
      assertThat(projected.op()).isEqualTo(bound.op());

      if (bound.isLiteralPredicate()) {
        assertThat(projected.literal().value())
            .isEqualTo(bound.asLiteralPredicate().literal().value());
      } else {
        assertThat(projected.literal()).isNull();
      }
    }
  }

  @Test
  public void testCaseSensitiveStrictIdentityProjection() {
    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).identity("id").build();
    Assertions.assertThatThrownBy(
            () -> Projections.strict(spec, true).project(Expressions.notNull("ID")))
        .isInstanceOf(ValidationException.class)
        .hasMessageContaining("Cannot find field 'ID' in struct");
  }

  @Test
  public void testBadSparkPartitionFilter() {
    // this tests a case that results in a full table scan in Spark with Hive tables. because the
    // hour field is not a partition, mixing it with partition columns in the filter expression
    // prevents the day/hour boundaries from being pushed to the metastore. this is an easy mistake
    // when tables are normally partitioned by both hour and dateint. the the filter is:
    //
    // WHERE dateint = 20180416
    //   OR (dateint = 20180415 and hour >= 20)
    //   OR (dateint = 20180417 and hour <= 4)

    Schema schema =
        new Schema(
            required(1, "id", Types.LongType.get()),
            optional(2, "data", Types.StringType.get()),
            required(3, "hour", Types.IntegerType.get()),
            required(4, "dateint", Types.IntegerType.get()));

    PartitionSpec spec = PartitionSpec.builderFor(schema).identity("dateint").build();

    Expression filter =
        or(
            equal("dateint", 20180416),
            or(
                and(equal("dateint", 20180415), greaterThanOrEqual("hour", 20)),
                and(equal("dateint", 20180417), lessThanOrEqual("hour", 4))));

    Expression projection = Projections.inclusive(spec).project(filter);

    Assertions.assertThat(projection).isInstanceOf(Or.class);
    Or or1 = (Or) projection;
    UnboundPredicate<?> dateint1 = assertAndUnwrapUnbound(or1.left());
    assertThat(dateint1.ref().name()).as("Should be a dateint predicate").isEqualTo("dateint");
    assertThat(dateint1.literal().value()).as("Should be dateint=20180416").isEqualTo(20180416);
    Assertions.assertThat(or1.right()).isInstanceOf(Or.class);
    Or or2 = (Or) or1.right();
    UnboundPredicate<?> dateint2 = assertAndUnwrapUnbound(or2.left());
    assertThat(dateint2.ref().name()).as("Should be a dateint predicate").isEqualTo("dateint");
    assertThat(dateint2.literal().value()).as("Should be dateint=20180415").isEqualTo(20180415);
    UnboundPredicate<?> dateint3 = assertAndUnwrapUnbound(or2.right());
    assertThat(dateint3.ref().name()).as("Should be a dateint predicate").isEqualTo("dateint");
    assertThat(dateint3.literal().value()).as("Should be dateint=20180417").isEqualTo(20180417);
  }

  @Test
  @SuppressWarnings("unchecked")
  public void testProjectionNames() {
    final Schema schema =
        new Schema(
            required(1, "timestamp1", Types.TimestampType.withoutZone()),
            optional(2, "timestamp2", Types.TimestampType.withoutZone()),
            optional(3, "timestamp3", Types.TimestampType.withoutZone()),
            optional(4, "timestamp4", Types.TimestampType.withoutZone()),
            optional(5, "date1", Types.DateType.get()),
            optional(6, "date2", Types.DateType.get()),
            optional(7, "date3", Types.DateType.get()),
            optional(8, "long", Types.LongType.get()),
            optional(9, "string", Types.StringType.get()));

    final PartitionSpec partitionSpec =
        PartitionSpec.builderFor(schema)
            .withSpecId(0)
            .hour("timestamp1")
            .day("timestamp2")
            .month("timestamp3")
            .year("timestamp4")
            .day("date1")
            .month("date2")
            .year("date3")
            .bucket("long", 10)
            .truncate("string", 10)
            .build();

    UnboundPredicate<Integer> predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(hour("timestamp1"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp1_hour");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(hour("timestamp1"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp1_hour");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(day("timestamp2"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp2_day");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(day("timestamp2"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp2_day");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(month("timestamp3"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp3_month");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(month("timestamp3"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp3_month");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(year("timestamp4"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp4_year");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(year("timestamp4"), 20));
    assertThat(predicate.ref().name()).isEqualTo("timestamp4_year");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(day("date1"), 20));
    assertThat(predicate.ref().name()).isEqualTo("date1_day");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(day("date1"), 20));
    assertThat(predicate.ref().name()).isEqualTo("date1_day");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(month("date2"), 20));
    assertThat(predicate.ref().name()).isEqualTo("date2_month");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(month("date2"), 20));
    assertThat(predicate.ref().name()).isEqualTo("date2_month");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(year("date3"), 20));
    assertThat(predicate.ref().name()).isEqualTo("date3_year");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(year("date3"), 20));
    assertThat(predicate.ref().name()).isEqualTo("date3_year");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(bucket("long", 10), 20));
    assertThat(predicate.ref().name()).isEqualTo("long_bucket");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(bucket("long", 10), 20));
    assertThat(predicate.ref().name()).isEqualTo("long_bucket");

    predicate =
        (UnboundPredicate<Integer>)
            Projections.strict(partitionSpec).project(equal(truncate("string", 10), "abc"));
    assertThat(predicate.ref().name()).isEqualTo("string_trunc");
    predicate =
        (UnboundPredicate<Integer>)
            Projections.inclusive(partitionSpec).project(equal(truncate("string", 10), "abc"));
    assertThat(predicate.ref().name()).isEqualTo("string_trunc");
  }
}

File: python/tests/expressions/test_projection.py
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.
# pylint:disable=redefined-outer-name

import pytest

from pyiceberg.expressions import (
    AlwaysTrue,
    And,
    EqualTo,
    GreaterThan,
    GreaterThanOrEqual,
    In,
    IsNull,
    LessThan,
    LessThanOrEqual,
    Not,
    NotEqualTo,
    NotIn,
    NotNull,
    Or,
)
from pyiceberg.expressions.visitors import inclusive_projection
from pyiceberg.partitioning import PartitionField, PartitionSpec
from pyiceberg.schema import Schema
from pyiceberg.transforms import (
    BucketTransform,
    DayTransform,
    HourTransform,
    IdentityTransform,
    TruncateTransform,
)
from pyiceberg.types import (
    DateType,
    LongType,
    NestedField,
    StringType,
    TimestampType,
)


@pytest.fixture
def schema() -> Schema:
    return Schema(
        NestedField(1, "id", LongType(), required=False),
        NestedField(2, "data", StringType(), required=False),
        NestedField(3, "event_date", DateType(), required=False),
        NestedField(4, "event_ts", TimestampType(), required=False),
    )


@pytest.fixture
def empty_spec() -> PartitionSpec:
    return PartitionSpec()


@pytest.fixture
def id_spec() -> PartitionSpec:
    return PartitionSpec(PartitionField(1, 1000, IdentityTransform(), "id_part"))


@pytest.fixture
def bucket_spec() -> PartitionSpec:
    return PartitionSpec(PartitionField(2, 1000, BucketTransform(16), "data_bucket"))


@pytest.fixture
def day_spec() -> PartitionSpec:
    return PartitionSpec(PartitionField(4, 1000, DayTransform(), "date"), PartitionField(3, 1000, DayTransform(), "ddate"))


@pytest.fixture
def hour_spec() -> PartitionSpec:
    return PartitionSpec(PartitionField(4, 1000, HourTransform(), "hour"))


@pytest.fixture
def truncate_str_spec() -> PartitionSpec:
    return PartitionSpec(PartitionField(2, 1000, TruncateTransform(2), "data_trunc"))


@pytest.fixture
def truncate_int_spec() -> PartitionSpec:
    return PartitionSpec(PartitionField(1, 1000, TruncateTransform(10), "id_trunc"))


@pytest.fixture
def id_and_bucket_spec() -> PartitionSpec:
    return PartitionSpec(
        PartitionField(1, 1000, IdentityTransform(), "id_part"), PartitionField(2, 1001, BucketTransform(16), "data_bucket")
    )


def test_identity_projection(schema: Schema, id_spec: PartitionSpec) -> None:
    predicates = [
        NotNull("id"),
        IsNull("id"),
        LessThan("id", 100),
        LessThanOrEqual("id", 101),
        GreaterThan("id", 102),
        GreaterThanOrEqual("id", 103),
        EqualTo("id", 104),
        NotEqualTo("id", 105),
        In("id", {3, 4, 5}),
        NotIn("id", {3, 4, 5}),
    ]

    expected = [
        NotNull("id_part"),
        IsNull("id_part"),
        LessThan("id_part", 100),
        LessThanOrEqual("id_part", 101),
        GreaterThan("id_part", 102),
        GreaterThanOrEqual("id_part", 103),
        EqualTo("id_part", 104),
        NotEqualTo("id_part", 105),
        In("id_part", {3, 4, 5}),
        NotIn("id_part", {3, 4, 5}),
    ]

    project = inclusive_projection(schema, id_spec)
    for index, predicate in enumerate(predicates):
        expr = project(predicate)
        assert expected[index] == expr


def test_bucket_projection(schema: Schema, bucket_spec: PartitionSpec) -> None:
    predicates = [
        NotNull("data"),
        IsNull("data"),
        LessThan("data", "val"),
        LessThanOrEqual("data", "val"),
        GreaterThan("data", "val"),
        GreaterThanOrEqual("data", "val"),
        EqualTo("data", "val"),
        NotEqualTo("data", "val"),
        In("data", {"v1", "v2", "v3"}),
        NotIn("data", {"v1", "v2", "v3"}),
    ]

    expected = [
        NotNull("data_bucket"),
        IsNull("data_bucket"),
        AlwaysTrue(),
        AlwaysTrue(),
        AlwaysTrue(),
        AlwaysTrue(),
        EqualTo("data_bucket", 14),
        AlwaysTrue(),
        In("data_bucket", {1, 3, 13}),
        AlwaysTrue(),
    ]

    project = inclusive_projection(schema, bucket_spec)
    for index, predicate in enumerate(predicates):
        expr = project(predicate)
        assert expected[index] == expr


def test_hour_projection(schema: Schema, hour_spec: PartitionSpec) -> None:
    predicates = [
        NotNull("event_ts"),
        IsNull("event_ts"),
        LessThan("event_ts", "2022-11-27T10:00:00"),
        LessThanOrEqual("event_ts", "2022-11-27T10:00:00"),
        GreaterThan("event_ts", "2022-11-27T09:59:59.999999"),
        GreaterThanOrEqual("event_ts", "2022-11-27T09:59:59.999999"),
        EqualTo("event_ts", "2022-11-27T10:00:00"),
        NotEqualTo("event_ts", "2022-11-27T10:00:00"),
        In("event_ts", {"2022-11-27T10:00:00", "2022-11-27T09:59:59.999999"}),
        NotIn("event_ts", {"2022-11-27T10:00:00", "2022-11-27T09:59:59.999999"}),
    ]

    expected = [
        NotNull("hour"),
        IsNull("hour"),
        LessThanOrEqual("hour", 463761),
        LessThanOrEqual("hour", 463762),
        GreaterThanOrEqual("hour", 463762),
        GreaterThanOrEqual("hour", 463761),
        EqualTo("hour", 463762),
        AlwaysTrue(),
        In("hour", {463761, 463762}),
        AlwaysTrue(),
    ]

    project = inclusive_projection(schema, hour_spec)
    for index, predicate in enumerate(predicates):
        expr = project(predicate)
        assert expected[index] == expr, predicate


def test_day_projection(schema: Schema, day_spec: PartitionSpec) -> None:
    predicates = [
        NotNull("event_ts"),
        IsNull("event_ts"),
        LessThan("event_ts", "2022-11-27T00:00:00"),
        LessThanOrEqual("event_ts", "2022-11-27T00:00:00"),
        GreaterThan("event_ts", "2022-11-26T23:59:59.999999"),
        GreaterThanOrEqual("event_ts", "2022-11-26T23:59:59.999999"),
        EqualTo("event_ts", "2022-11-27T10:00:00"),
        NotEqualTo("event_ts", "2022-11-27T10:00:00"),
        In("event_ts", {"2022-11-27T00:00:00", "2022-11-26T23:59:59.999999"}),
        NotIn("event_ts", {"2022-11-27T00:00:00", "2022-11-26T23:59:59.999999"}),
    ]

    expected = [
        NotNull("date"),
        IsNull("date"),
        LessThanOrEqual("date", 19322),
        LessThanOrEqual("date", 19323),
        GreaterThanOrEqual("date", 19323),
        GreaterThanOrEqual("date", 19322),
        EqualTo("date", 19323),
        AlwaysTrue(),
        In("date", {19322, 19323}),
        AlwaysTrue(),
    ]

    project = inclusive_projection(schema, day_spec)
    for index, predicate in enumerate(predicates):
        expr = project(predicate)
        assert expected[index] == expr, predicate


def test_date_day_projection(schema: Schema, day_spec: PartitionSpec) -> None:
    predicates = [
        NotNull("event_date"),
        IsNull("event_date"),
        LessThan("event_date", "2022-11-27"),
        LessThanOrEqual("event_date", "2022-11-27"),
        GreaterThan("event_date", "2022-11-26"),
        GreaterThanOrEqual("event_date", "2022-11-26"),
        EqualTo("event_date", "2022-11-27"),
        NotEqualTo("event_date", "2022-11-27"),
        In("event_date", {"2022-11-26", "2022-11-27"}),
        NotIn("event_date", {"2022-11-26", "2022-11-27"}),
    ]

    expected = [
        NotNull("ddate"),
        IsNull("ddate"),
        LessThanOrEqual("ddate", 19322),
        LessThanOrEqual("ddate", 19323),
        GreaterThanOrEqual("ddate", 19323),
        GreaterThanOrEqual("ddate", 19322),
        EqualTo("ddate", 19323),
        AlwaysTrue(),
        In("ddate", {19322, 19323}),
        AlwaysTrue(),
    ]

    project = inclusive_projection(schema, day_spec)
    for index, predicate in enumerate(predicates):
        expr = project(predicate)
        assert expected[index] == expr, predicate


def test_string_truncate_projection(schema: Schema, truncate_str_spec: PartitionSpec) -> None:
    predicates = [
        NotNull("data"),
        IsNull("data"),
        LessThan("data", "aaa"),
        LessThanOrEqual("data", "aaa"),
        GreaterThan("data", "aaa"),
        GreaterThanOrEqual("data", "aaa"),
        EqualTo("data", "aaa"),
        NotEqualTo("data", "aaa"),
        In("data", {"aaa", "aab"}),
        NotIn("data", {"aaa", "aab"}),
    ]

    expected = [
        NotNull("data_trunc"),
        IsNull("data_trunc"),
        LessThanOrEqual("data_trunc", "aa"),
        LessThanOrEqual("data_trunc", "aa"),
        GreaterThanOrEqual("data_trunc", "aa"),
        GreaterThanOrEqual("data_trunc", "aa"),
        EqualTo("data_trunc", "aa"),
        AlwaysTrue(),
        EqualTo("data_trunc", "aa"),
        AlwaysTrue(),
    ]

    project = inclusive_projection(schema, truncate_str_spec)
    for index, predicate in enumerate(predicates):
        expr = project(predicate)
        assert expected[index] == expr, predicate


def test_int_truncate_projection(schema: Schema, truncate_int_spec: PartitionSpec) -> None:
    predicates = [
        NotNull("id"),
        IsNull("id"),
        LessThan("id", 10),
        LessThanOrEqual("id", 10),
        GreaterThan("id", 9),
        GreaterThanOrEqual("id", 10),
        EqualTo("id", 15),
        NotEqualTo("id", 15),
        In("id", {15, 16}),
        NotIn("id", {15, 16}),
    ]

    expected = [
        NotNull("id_trunc"),
        IsNull("id_trunc"),
        LessThanOrEqual("id_trunc", 0),
        LessThanOrEqual("id_trunc", 10),
        GreaterThanOrEqual("id_trunc", 10),
        GreaterThanOrEqual("id_trunc", 10),
        EqualTo("id_trunc", 10),
        AlwaysTrue(),
        EqualTo("id_trunc", 10),
        AlwaysTrue(),
    ]

    project = inclusive_projection(schema, truncate_int_spec)
    for index, predicate in enumerate(predicates):
        expr = project(predicate)
        assert expected[index] == expr, predicate


def test_projection_case_sensitive(schema: Schema, id_spec: PartitionSpec) -> None:
    project = inclusive_projection(schema, id_spec)
    with pytest.raises(ValueError) as exc_info:
        project(NotNull("ID"))
        assert str(exc_info) == "Could not find field with name ID, case_sensitive=True"


def test_projection_case_insensitive(schema: Schema, id_spec: PartitionSpec) -> None:
    project = inclusive_projection(schema, id_spec, case_sensitive=False)
    assert NotNull("id_part") == project(NotNull("ID"))


def test_projection_empty_spec(schema: Schema, empty_spec: PartitionSpec) -> None:
    project = inclusive_projection(schema, empty_spec)
    assert AlwaysTrue() == project(And(LessThan("id", 5), NotNull("data")))


def test_and_projection_multiple_projected_fields(schema: Schema, id_and_bucket_spec: PartitionSpec) -> None:
    project = inclusive_projection(schema, id_and_bucket_spec)
    assert project(And(LessThan("id", 5), In("data", {"a", "b", "c"}))) == And(
        LessThan("id_part", 5), In("data_bucket", {2, 3, 15})
    )


def test_or_projection_multiple_projected_fields(schema: Schema, id_and_bucket_spec: PartitionSpec) -> None:
    project = inclusive_projection(schema, id_and_bucket_spec)
    assert project(Or(LessThan("id", 5), In("data", {"a", "b", "c"}))) == Or(
        LessThan("id_part", 5), In("data_bucket", {2, 3, 15})
    )


def test_not_projection_multiple_projected_fields(schema: Schema, id_and_bucket_spec: PartitionSpec) -> None:
    project = inclusive_projection(schema, id_and_bucket_spec)
    # Not causes In to be rewritten to NotIn, which cannot be projected
    assert project(Not(Or(LessThan("id", 5), In("data", {"a", "b", "c"})))) == GreaterThanOrEqual("id_part", 5)


def test_projection_partial_projected_fields(schema: Schema, id_spec: PartitionSpec) -> None:
    project = inclusive_projection(schema, id_spec)
    assert project(And(LessThan("id", 5), In("data", {"a", "b", "c"}))) == LessThan("id_part", 5)
